assert.Eventually with low ticks cause panic: send on closed channel

# Setup
If you put an:
`assert.Eventually(t, func() bool { return true }, time.Millisecond, 50 * time.Microsecond)` inside your test, you will get flaky results and more often then not, you will get a panic:
`panic: send on closed channel`

# Workaround
Increasing the tick to around `100 * time.Microsecond` removes the flakyness

# Solution
Add a waitgroup potentially so that [this channel close](https://github.com/stretchr/testify/blob/221dbe5ed46703ee255b1da0dec05086f5035f62/assert/assertions.go#L1483) waits for the gofunc writing to the channel completes.

Same for me. To reproduce:
> `go version`
```
go version go1.13 darwin/amd64
```

> `go.mod`
```
module reqtest

go 1.13

require github.com/stretchr/testify v1.4.0
```

> `req_test.go`
```
package main

import (
	"testing"
	"time"

	"github.com/stretchr/testify/require"
)

func Test_Eventually(t *testing.T) {
	require.Eventually(t, func() bool { return true }, time.Millisecond, time.Microsecond)
}
```

> `go test req_test.go`
```
--- FAIL: Test_Eventually (0.00s)
panic: send on closed channel
````

--- a/assert/assertions.go
+++ b/assert/assertions.go
@@ -1475,24 +1475,26 @@ func Eventually(t TestingT, condition func() bool, waitFor time.Duration, tick t
 		h.Helper()
 	}
 
+	ch := make(chan bool, 1)
+
 	timer := time.NewTimer(waitFor)
-	ticker := time.NewTicker(tick)
-	checkPassed := make(chan bool)
 	defer timer.Stop()
+
+	ticker := time.NewTicker(tick)
 	defer ticker.Stop()
-	defer close(checkPassed)
-	for {
+
+	for tick := ticker.C; ; {
 		select {
 		case <-timer.C:
-			return Fail(t, "Condition never satisfied", msgAndArgs...)
-		case result := <-checkPassed:
-			if result {
+			return false
+		case <-tick:
+			tick = nil
+			go func() { ch <- condition() }()
+		case v := <-ch:
+			if v {
 				return true
 			}
-		case <-ticker.C:
-			go func() {
-				checkPassed <- condition()
-			}()
+			tick = ticker.C
 		}
 	}
 }

--- a/assert/assertions_test.go
+++ b/assert/assertions_test.go
@@ -1921,3 +1921,12 @@ func TestEventuallyTrue(t *testing.T) {
 
 	True(t, Eventually(t, condition, 100*time.Millisecond, 20*time.Millisecond))
 }
+
+func TestEventuallyIssue805(t *testing.T) {
+	mockT := new(testing.T)
+
+	NotPanics(t, func() {
+		condition := func() bool { <-time.After(time.Millisecond); return true }
+		False(t, Eventually(mockT, condition, time.Millisecond, time.Microsecond))
+	})
+}

